axum-extra/src/response/file_stream.rs

```diff
@@ -126,6 +133,165 @@ where
+    ///
+    ///     FileStream::<AsyncReaderStream>::try_range_response(Path::new("CHANGELOG.md"),range_start,range_end,buffer_size).await
+    ///     .map_err(|e| (StatusCode::NOT_FOUND, format!("File not found: {e}")))
+    ///     .into_response()
+    ///     
+    /// }
+    /// let app = Router::new().route("/FileStreamRange", get(range_stream));
+    /// # let _: Router = app;
+    /// ```
+    pub async fn try_range_response(
+        file_path: &Path,
+        start: u64,
+        mut end: u64,
+        buffer_size: usize,
+    ) -> io::Result<Response> {
+        // open file
+        let file = File::open(file_path).await?;
+
+        // get file metadata
+        let metadata = file.metadata().await?;
+        let total_size = metadata.len();
+
+        if end == 0 {
+            end = total_size - 1;
+        }
+
+        // range check
+        if start > total_size {
+            return Ok((StatusCode::RANGE_NOT_SATISFIABLE, "Range Not Satisfiable").into_response());
+        }
+        if start > end {
+            return Ok((StatusCode::RANGE_NOT_SATISFIABLE, "Range Not Satisfiable").into_response());
+        }
+        if end >= total_size {
+            return Ok((StatusCode::RANGE_NOT_SATISFIABLE, "Range Not Satisfiable").into_response());
+        }
+
+        // get file stream
+        let stream = try_stream(file, start, end, buffer_size).await?;
+        let mut resp = Response::builder().header(header::CONTENT_TYPE, "application/octet-stream");
+        resp = resp.status(StatusCode::PARTIAL_CONTENT);
+
+        resp = resp.header(
+            header::CONTENT_RANGE,
+            format!("bytes {}-{}/{}", start, end, total_size),
+        );
+
+        Ok(resp
+            .body(body::Body::from_stream(stream))
+            .unwrap_or_else(|e| {
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    format!("build FileStream responsec error: {}", e),
+                )
+                    .into_response()
+            }))
+    }
+}
+
+/// More complex manipulation of files and conversion to a stream
+async fn try_stream(
+    mut file: File,
+    start: u64,
+    end: u64,
+    buffer_size: usize,
+) -> Result<impl Stream<Item = Result<Vec<u8>, io::Error>>, io::Error> {
+    file.seek(std::io::SeekFrom::Start(start)).await?;
+
+    let mut buffer = vec![0; buffer_size];
+
+    let stream = async_stream::try_stream! {
+        let mut total_read = 0;
+
+            while total_read < end {
+                let bytes_to_read = std::cmp::min(buffer_size as u64, end - total_read);
+                let n = file.read(&mut buffer[..bytes_to_read as usize]).await.map_err(|e| {
+                    std::io::Error::new(std::io::ErrorKind::Other, e)
+                })?;
+                if n == 0 {
+                    break; // EOF
+                }
+                total_read += n as u64;
+                yield buffer[..n].to_vec();
+
+        }
+    };
+    Ok(stream)
```

Comment:
It should be possible to replace this entire part with [`AsyncReadExt::take`](https://docs.rs/tokio/1.42.0/tokio/io/trait.AsyncReadExt.html#method.take)

Reply:
> AsyncReadExt::take

Yes, it's better to use `AsyncReadExt::take `replacement here, I fixed the

Reply:
> It should be possible to replace this entire part with [`AsyncReadExt::take`](https://docs.rs/tokio/1.42.0/tokio/io/trait.AsyncReadExt.html#method.take) and then going back to `ReaderStream`

Yes, it's better to use AsyncReadExt::take replacement here, I fixed the